src/core/shards/logging.cpp

```diff
@@ -100,30 +103,166 @@ struct Msg : public LoggingBase {
   }
 };
 
-// Register Log
-RUNTIME_CORE_SHARD(Log);
-RUNTIME_SHARD_help(Log);
-RUNTIME_SHARD_inputTypes(Log);
-RUNTIME_SHARD_outputTypes(Log);
-RUNTIME_SHARD_parameters(Log);
-RUNTIME_SHARD_setParam(Log);
-RUNTIME_SHARD_getParam(Log);
-RUNTIME_SHARD_activate(Log);
-RUNTIME_SHARD_END(Log);
-
-// Register Msg
-RUNTIME_CORE_SHARD(Msg);
-RUNTIME_SHARD_help(Msg);
-RUNTIME_SHARD_inputTypes(Msg);
-RUNTIME_SHARD_outputTypes(Msg);
-RUNTIME_SHARD_parameters(Msg);
-RUNTIME_SHARD_setParam(Msg);
-RUNTIME_SHARD_getParam(Msg);
-RUNTIME_SHARD_activate(Msg);
-RUNTIME_SHARD_END(Msg);
+/*
+ * Custom ring buffer sink
+ */
+template <typename Mutex> class custom_ringbuffer_sink final : public spdlog::sinks::base_sink<Mutex> {
+public:
+  explicit custom_ringbuffer_sink(size_t n_items, spdlog::level::level_enum min_level = spdlog::level::debug)
+      : q_{n_items}, min_level_(min_level) {}
+
+  inline size_t generation() { return _generation; }
+
+  inline bool is_dirty() { return _last_generation < _generation; }
+
+  std::vector<std::string> last_formatted() {
+    std::lock_guard<Mutex> lock(spdlog::sinks::base_sink<Mutex>::mutex_);
+
+    if (is_dirty()) {
```

Comment:
can't dirty be atomic so to avoid locking at all?

Reply:
We need locking because `q_` could otherwise be written to. I should move the lock inside the branch maybe.

Though you are right that `is_dirty()` isn't atomic so the lock might also be needed for it. I could add a `bool` that is recomputed when necessary (same as previous commit), which makes it atomic at the expense of one additional field.

Reply:
what I mean is lock only if dirty basically, as dirty can be an atomic bool

Reply:
Use a simple boolean for now which should be atomic.